tactics/tactics.ml

```diff
@@ -2096,60 +2100,62 @@ let check_decl env sigma decl =
     let env = Proofview.Goal.env gl in
     let concl = Proofview.Goal.concl gl in
     let sigma = Tacmach.project gl in
     let ctx = named_context env in
-    let map = function
-    | LocalAssum (id,t) as decl ->
-      let () = if List.mem_f Id.equal id.binder_name ids then
-        error (VariableHasNoValue id.binder_name);
-      in
-      decl
-    | LocalDef (id,_,t) as decl ->
-      if List.mem_f Id.equal id.binder_name ids then LocalAssum (id, t) else decl
-    in
-    let ctx = List.map map ctx in
-    let base_env = reset_context env in
-    let env = push_named_context ctx base_env in
-    let check =
-      try
-        let check (env, sigma, seen) decl =
-          (* Do no recheck hypotheses that do not depend *)
-          let sigma =
-            if not seen then sigma
-            else if List.exists (fun id -> occur_var_in_decl env sigma id decl) ids then
-              check_decl env sigma decl
-            else sigma
+    let ids = Id.Set.of_list idl in
+    (* We assume the context to respect dependencies *)
+    let rec fold ids ctx =
+      if Id.Set.is_empty ids then
+        let base_env = reset_context env in
+        let env = push_named_context ctx base_env in
+        env, sigma, Id.Set.empty
+      else
+        match ctx with
+        | [] -> assert false
+        | decl :: ctx ->
+          let decl, ids, found =
+            match decl with
+            | LocalAssum (id,t) ->
+              let () =
+                if Id.Set.mem id.binder_name ids then
+                  error (VariableHasNoValue id.binder_name)
+              in
+              decl, ids, false
+            | LocalDef (id,_,t) as decl ->
+               if Id.Set.mem id.binder_name ids
+               then LocalAssum (id, t), Id.Set.remove id.binder_name ids, true
+               else decl, ids, false
           in
-          let seen = seen || List.mem_f Id.equal (NamedDecl.get_id decl) ids in
-          (push_named decl env, sigma, seen)
-        in
-        let (env, sigma, _) = List.fold_left check (base_env, sigma, false) (List.rev ctx) in
-        let sigma =
-          if List.exists (fun id -> occur_var env sigma id concl) ids then
-            check_is_type env sigma concl
-          else sigma
-        in
-        Proofview.Unsafe.tclEVARS sigma
-      with DependsOnBody where as exn ->
-        let _, info = Exninfo.capture exn in
-        let msg = match where with
-        | None -> str "Conclusion" ++ on_the_bodies ids
-        | Some id -> str "Hypothesis " ++ Id.print id ++ on_the_bodies ids
-        in
-        Tacticals.tclZEROMSG ~info msg
+          let env, sigma, ids = fold ids ctx in
+          if Id.Set.exists (fun id -> occur_var_in_decl env sigma id decl) ids then
+            let sigma = check_decl env sigma idl ids decl in (* can sigma really change? *)
+            let ids = Id.Set.add (get_id decl) ids in
+            push_named decl env, sigma, Id.Set.add (get_id decl) ids
```

Comment:
Is there a reason that you add `get_id decl` to `ids` on both this line and the previous one?

Reply:
Some overlook. Shall we do a quick PR to remove the confusion?